backends/dpdk/dpdkArch.cpp

```diff
@@ -2448,4 +2448,242 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
+                        "P4Runtime requires that Ternary matches have masked-off bits set to 0, "
+                        "updating value %1% to conform to the P4Runtime specification", km->left);
+            value &= mask;
+        }
+        valueStr = value;
+        maskStr = mask;
+    } else {
+        valueStr = convertSimpleKeyExpressionToBigInt(k, keyWidth, typeMap);
+        maskStr = Util::mask(keyWidth);
+    }
+    print(valueStr, "/");
+    print(maskStr, " ");
+}
+
+void EmitDpdkTableConfig::addRange(const IR::Expression* k, int keyWidth, P4::TypeMap* typeMap) {
+    if (k->is<IR::DefaultExpression>())  // don't care, skip in P4Runtime message
+        return;
+    big_int startStr;
+    big_int endStr;
+    if (k->is<IR::Range>()) {
+        auto kr = k->to<IR::Range>();
+        auto start = convertSimpleKeyExpressionToBigInt(kr->left, keyWidth, typeMap);
+        auto end = convertSimpleKeyExpressionToBigInt(kr->right, keyWidth, typeMap);
+        // Error on invalid range values
+        big_int maxValue = (big_int(1) << keyWidth) - 1;
+        // NOTE: If end value is > max allowed for keyWidth, value gets
+        // wrapped around. A warning is issued in this case by the frontend
+        // earlier.
+        // For e.g. 16 bit key has a max value of 65535, Range of (1..65536)
+        // will be converted to (1..0) and will fail below check.
+        if (start > end)
+            ::error(ErrorType::ERR_INVALID, "%s Invalid range for table entry", kr->srcInfo);
+        startStr = start;
+        endStr = end;
+    } else {
+        startStr = convertSimpleKeyExpressionToBigInt(k, keyWidth, typeMap);
+        endStr = startStr;
+    }
+    print(startStr,"/");
+    print(endStr,"");
+}
+
+void EmitDpdkTableConfig::addOptional(const IR::Expression* k, int keyWidth,
+                    P4::TypeMap* typeMap) {
+    if (k->is<IR::DefaultExpression>())  // don't care, skip in P4Runtime message
+        return;
+    auto value = convertSimpleKeyExpressionToBigInt(k, keyWidth, typeMap);
+    print(value," ");
+}
+
+void EmitDpdkTableConfig::addMatchKey(const IR::P4Table* table,
+                     const IR::ListExpression* keyset,
+                     P4::TypeMap* typeMap) {
+    int keyIndex = 0;
+    for (auto k : keyset->components) {
+            auto tableKey = table->getKey()->keyElements.at(keyIndex++);
+            auto keyWidth = getTypeWidth(tableKey->expression->type, typeMap);
+            auto matchType = getKeyMatchType(tableKey, refMap);
+            if (matchType == P4::P4CoreLibrary::instance.exactMatch.name) {
+              addExact(k, keyWidth, typeMap);
+            } else if (matchType == P4::P4CoreLibrary::instance.lpmMatch.name) {
+              addLpm(k, keyWidth, typeMap);
+            } else if (matchType == P4::P4CoreLibrary::instance.ternaryMatch.name) {
+              addTernary(k, keyWidth, typeMap);
+            } else if (matchType == P4V1::V1Model::instance.rangeMatchType.name) {
```

Comment:
The range and optional match type should also be available in PSA. Do we not have such definitions for PSA? I think we should avoid using the v1model.h

Reply:
done